### ISSUE TITLE
Deadlock before logfile banner while starting 100 instances of Appium server.

### ISSUE THREAD

**Issue by sleekweasel** [appium/appium#12833]
## The problem

Deadlock before logfile banner while starting 100 instances of Appium server.

At least, it looks like a deadlock - it lasts at least 90s when our watchdog thread kills it.

## Environment

* Appium version 1.11.1
* Same version does not exhibit the issue on other machines.
* Docker container running Linux automation2 4.14.75-0.2-default #1 SMP PREEMPT Mon Nov 13 14:53:06 UTC 2017 (c152297) x86_64 x86_64 x86_64 GNU/Linux 
* v8.4.0
* Npm 5.4.1
* Android
* emulator
* Appium CLI

## Details

Our CI starts 100 or so Appium servers - one for each emulator - to keep tracking issues through the logs easy.

This has been working well, but suddenly one of our CI machines has been experiencing deadlocks (with all 100 or none of the servers affected) at about 50% of each build-test cycle, while the Appium servers loads their modules, before the logfile is opened. In the last day, it's become 90% of build-test cycles.

All the server stack traces for each run fall into a handful of categories that end at the same place (though different for each run), suggesting that they manage to follow different routes through the require tree, with at least two processes mutually locking some files along the way which prevents all processes progressing.

I have lsof output for each server process, but the file each process holds looks to correlate with the stack trace group each falls into - I haven't verified that yet.

## Link to Appium logs

There are no logs generated: this occurs before the logs are opened, but I did manage to attach debuggers to each server. I'll upload a few candidates from one run's stack traces. There's only 1 unique stack trace, which I presume is the cause of the mess - w11. The others are all variously duplicated.

https://gist.github.com/sleekweasel/e6881067ad7209764374b7514a298b0a

## Code To Reproduce Issue [ Good To Have ]

There's no real code: I'm simply starting the server 100 times from 100 different shell processes, and this only fails on one particular machine.

echo Waiting for device ; 
timeout 3 adb wait-for-device || echo Wait failed. Continuing anyway. ; 
echo Starting appium command ; 
node --inspect-port=0 ./node_modules/.bin/appium --port "0" --chromedriver-port "0" --log-level warn:debug --suppress-adb-kill-server --log /local/buildAgent/work/dcce385aab59e3da/bma-calabash/staging-log/appium_85.log

obviously, the logfiles are numbered for each worker index.

I haven't tried the obvious intrusive things like restarting the Docker container, clearing the npm cache, etc, because it was such an odd problem it seemed better to explore it for a while.

**Comment by mykola-mokhnach** [comment]
I'd start with Appium and NodeJS/npm updates. Unfortunately, we don't provide patches to older versions

**Comment by sleekweasel** [comment]
@mykola-mokhnach Sure - as I say at the end I'm going to try resetting things and perhaps updating too (although updating is a full task in its own right with various incompatibilities to fix) - it's just that this behaviour hints at a deadlock condition I wouldn't expect to happen during module inclusion, so it looked interesting and potentially useful to have documented.

**Comment by mykola-mokhnach** [comment]
I suppose this might be an issue of graceful-fs module itself. That is why it is necessary to make sure it still happens with the most recent server version

**Comment by sleekweasel** [comment]
Small update - I changed node_modules into a softlink pointing to a directory that wasn't on a shared Docker volume, and the issue hasn't recurred. I'm concluding that it's some obscure Docker/overlayfs/kernel issue: all the stuck appium servers were in disk wait, and there were no file locks.

**Comment by sleekweasel** [comment]
We recently updated our servers to the last version of CentOS, so I tried running on shared storage again.

This is still happening with "appium": "v1.19.1-rc.1" with node_modules living in /root/ (I know, bad idea) in the container overlay - so not a shared volume even.

If I copy the files from /root/ to /tmp/node_modules and use them there, it's fine, although we suffer the time penalty of that operation.
